L74-java-channel-creds.md

```diff
@@ -0,0 +1,571 @@
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
+    .build();
+```
+
+`newChannelBuilder()` will iterate through the `ManagedChannelProviders`
+trying to create a builder for each in turn. The first provider that succeeds
+will be used. If no provider is able to handle the credentials, then
+`newChannelBuilder()` will throw. `newChannelBuilderForAddress()` will be a
+convenience function that creates the target string and calls
+`newChannelBuilder()`, mirroring its current behavior.
+
+The new ChannelCredentials API cannot be mixed with the pre-existing
+`useTransportSecurity()` and `usePlaintext()` builder APIs. The old-style
+methods will throw if called on a builder with a ChannelCredential provided.
+
+Note that this new API does not allow: 1) creating a channel without a
+credential and supplying the credential later and 2) changing the credential
+after creating the builder. This is not expected to be much of an issue as
+users typically have the security information available before creating the
+builder and reusing a builder is rare, especially to change the security.
+
+```java
+package io.grpc;
+
+/**
+ * Represents a security configuration to be used for channels. There is no
+ * generic mechanism for processing arbitrary ChannelCredentials; the consumer
+ * of the credential (the channel) must support each implementation explicitly
+ * and separately. Consumers are not required to support all types or even all
+ * possible configurations for types that are partially supported, but they
+ * <em>must</em> at least fully support ChoiceChannelCredentials.
+ *
+ * <p>A ChannelCredential provides client identity and authenticates the server.
+ * This is different from CallCredentials, which only provides client identity.
+ * They can also influence types of encryption used and similar security
+ * configuration.
+ *
+ * The concrete credential type should not be relevant to most users of the API
+ * and may be an implementation decision. Users should generally use the {@code
+ * ChannelCredentials} type for variables instead of the concrete type.
+ * Freshly-constructed credentials should be returned as {@code
+ * ChannelCredentials} instead of a concrete type to encourage this pattern.
+ * Concrete types would only be used after {@code instanceof} checks (which must
+ * consider {@code ChoiceChannelCredentials}!).
+ */
+public abstract class ChannelCredentials {}
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  public static ChannelCredentials create() {...}
+
+  /* Complicated. Described separately below. */
+}
+
+/** No client identity, authentication, or encryption is to be used. */
+public final class InsecureChannelCredentials extends ChannelCredentials {
```

Comment:
Sorry if I have asked this before. Is there a plan to support `LocalCredentials` like grpc-core/C++ does https://github.com/grpc/grpc/blob/master/include/grpcpp/security/credentials.h#L311 ? I think local creds can be used only for "local" connections such as UDS or "localhost".

Reply:
I still have no plan to support `LocalCredentials`. Java still doesn't support UDS and the use of "localhost" isn't near as important for Java as it is in C++, as we have in-process transport for testing.

We may end up having it later to reduce the number of cases that have to use "unsafe" APIs. Basically, every instance of `InsecureChannelCredentials` may be a security issue so it may be useful to have `LocalCredentials` which reduces the risk (it isn't _safe_, but it is less bad) and have fewer instances of `InsecureChannelCredentials` to continually monitor. But right now Java internally has very few usages of `usePlaintext()` because our testing story has been so different and UDS is a PITA in Java.

Reply:
Sounds good.